### PR TITLE
ci-secret-generator: support Vault

### PR DESCRIPTION
**Pull request by bbguimaraes**

### CODE REVIEW

**Comment by bbguimaraes** [comment]
Trying again with the fixes previously applied to the bootstrapper/generator.

`ci-secret-generator` now uses the shared client from https://github.com/openshift/ci-tools/pull/1826.

/hold cancel

https://issues.redhat.com/browse/DPTP-1974

File: pkg/secrets/flags.go [hunk 1]
```diff
@@ -71,12 +73,25 @@ func (o *CLIOptions) Complete(censor *DynamicCensor) error {
...
+	return o.NewClient(censor)
+}
+
 func (o *CLIOptions) NewClient(censor *DynamicCensor) (Client, error) {
 	if o.BwUser != "" {
 		c, err := bitwarden.NewClient(o.BwUser, o.BwPassword, censor.AddSecrets)
 		if err != nil {
 			return nil, fmt.Errorf("Failed to get Bitwarden client: %w", err)
 		}
+		if o.BwDefaultOrganization != "" || len(o.BwDefaultCollections) != 0 {
```
Comment (alvaroaleman) [inline pkg/secrets/flags.go:86]: Any reason to not hardcode that here? Do we ever use something different?

File: pkg/secrets/vault.go [hunk 2]
```diff
@@ -55,6 +56,22 @@ func (c *vaultClient) getSecretAtPath(path, key string) ([]byte, error) {
...
+	var data map[string]string
+	if current, err := c.upstream.GetKV(path); err != nil {
+		if !vaultclient.IsNotFound(err) {
+			return err
+		}
+		data = map[string]string{field: content}
+	} else {
+		data = current.Data
+		data[field] = content
+	}
```
Comment (bbguimaraes) [inline pkg/secrets/vault.go:70]: I forgot to mention: this is ugly, but required to conform to the current client interface based on the MO of the Bitwarden client.

I decided not to do anything too clever until we have some measurement of the overhead (or remove the old Bitwarden client so a more fitting interface can be used).

File: pkg/secrets/flags.go (hunk 1 above)
Comment (bbguimaraes) [reply pkg/secrets/flags.go]: No real reason, https://github.com/openshift/ci-tools/pull/1779/commits/e0eb3465d05b23921dfa762ba43378bed08caae5 makes it always apply.

File: cmd/ci-secret-generator/main.go [hunk 3]
```diff
@@ -64,17 +55,17 @@ type fieldGenerator struct {
...
-	flag.CommandLine.StringVar(&o.bootstrapConfigPath, "bootstrap-config", "", "Path to the config file used for bootstrapping cluster secrets after using this tool.")
-	flag.CommandLine.BoolVar(&o.validate, "validate", true, "Validate that the items created from this tool are used in bootstrapping")
-	flag.CommandLine.BoolVar(&o.validateOnly, "validate-only", false, "If the tool should exit after the validation")
-	flag.CommandLine.StringVar(&o.outputFile, "output-file", "", "output file for dry-run mode")
-	flag.CommandLine.StringVar(&o.bwUser, "bw-user", "", "Username to access BitWarden.")
-	flag.CommandLine.StringVar(&o.bwPasswordPath, "bw-password-path", "", "Path to a password file to access BitWarden.")
-	flag.CommandLine.StringVar(&o.logLevel, "log-level", "info", fmt.Sprintf("Log level is one of %v.", logrus.AllLevels))
-	flag.CommandLine.IntVar(&o.maxConcurrency, "concurrency", 1, "Maximum number of concurrent in-flight goroutines to BitWarden.")
-	if err := flag.CommandLine.Parse(os.Args[1:]); err != nil {
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
```
Comment (alvaroaleman) [inline cmd/ci-secret-generator/main.go:58]: Those two are the exact same thing https://golang.org/pkg/flag/#pkg-variables :)

**Review by alvaroaleman — APPROVED** [review]

**Pull request merged**